src/suggestion.rs

```diff
@@ -50,14 +69,204 @@ impl fmt::Display for Detector {
+        let marker_size = head_sub_range.len() + DOTS_LEN + tail_sub_range.len();
+
+        (marker_size, shortened)
+    } else {
+        let full: String = stripped_line
+            .chars()
+            .skip(mistake_range.start)
+            .take(mistake_range.len())
+            .collect();
+        (marker_size, full)
+    };
+
+    let stripped_line_len = stripped_line.chars().count();
+
+    // full, uncut context coverage
+    let left_context = Range {
+        start: 0,
+        end: mistake_range.start,
+    };
+    let right_context = Range {
+        start: mistake_range.end,
+        end: stripped_line_len,
+    };
+
+    let avail_space = terminal_size
+        .saturating_sub(marker_size + padding_till_excerpt_start + TOTAL_CONTEXT_CHAR_COUNT);
+
+    // left and right we would like to partition the remaining space equally
+    let avail_space_half = avail_space / 2usize;
+
+    // @todo introduce a threshold, so the shortened version is not longer than than the original
+    let (left_context, right_context) = match (
+        avail_space_half > left_context.len(),
+        avail_space_half > right_context.len(),
+    ) {
+        (true, false) => {
+            // left context does not use all the capacity avail
+            // allow the right context to consume the excess.
+            let right_avail_space = avail_space - left_context.len();
+            (
+                left_context,
+                Range {
+                    start: right_context.end,
+                    end: cmp::min(mistake_range.end + right_avail_space, stripped_line_len),
+                },
+            )
+        }
+        (false, true) => {
+            // right context does not use all the capacity avail
+            // allow the left context to consume the excess.
+            let left_avail_space = avail_space - right_context.len();
+            (
+                Range {
+                    start: left_context.end.saturating_sub(left_avail_space),
+                    end: left_context.end,
+                },
+                right_context,
+            )
+        }
+        (false, false) => {
+            // both sides have excess chars, so yield `avail_space_half` to both sides
+            (
+                Range {
+                    start: left_context.end.saturating_sub(avail_space_half),
+                    end: left_context.end,
+                },
+                Range {
+                    start: right_context.start,
+                    end: right_context.start + avail_space_half,
+                },
+            )
+        }
+        _ => {
+            // both sides are less than the allowed context, no need to modify
+            (left_context, right_context)
+        }
+    };
+
+    assert!(left_context.end == mistake_range.start);
+    assert!(right_context.end <= stripped_line_len);
+    assert!(left_context.len() + mistake_range.len() + right_context.len() <= stripped_line_len);
```

Comment:
Are these `assert`s on purpose or are `debug_assert`s maybe better suited here?

Reply:
Puuh, frankly I am not sure - the requirements here should really be true always, but the remaining program could very well continue operation if they were not true, so I guess `debug_assert!` would be better.